File: 0450/recursive.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        if root is None:
            return root
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        elif key < root.val:
            root.left = self.deleteNode(root.left, key)
        else:  # key == root.val
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            node = self.findMinNode(root.right)
            root.val = node.val
            root.right = self.deleteNode(root.right, root.val)
        return root

    def findMinNode(self, node):
        while node.left is not None:
            node = node.left
        return node


if __name__ == '__main__':
    sol = Solution()

    root = TreeNode(5)
    root.left = TreeNode(3)
    root.right = TreeNode(6)
    root.left.left = TreeNode(2)
    root.left.right = TreeNode(4)
    root.right.right = TreeNode(7)
    key = 3
    got = sol.deleteNode(root, key)
    assert got.val == 5
    assert got.left.val == 4
    assert got.right.val == 6
    assert got.left.left.val == 2
    assert got.right.right.val == 7

    root = TreeNode(5)
    root.left = TreeNode(3)
    root.right = TreeNode(6)
    root.left.left = TreeNode(2)
    root.left.right = TreeNode(4)
    root.right.right = TreeNode(7)
    key = 0
    got = sol.deleteNode(root, key)
    assert got.val == 5
    assert got.left.val == 3
    assert got.right.val == 6
    assert got.left.left.val == 2
    assert got.left.right.val == 4
    assert got.right.right.val == 7

File: 0450/recursive.cpp
#include <cstdlib>
#include <string>

#include "gtest/gtest.h"
#include "leetcode/tree.hpp"

using namespace std;
using namespace LeetCode;

class Solution {
 public:
  TreeNode* deleteNode(TreeNode* root, int key) {
    if (!root) {
      return nullptr;
    } else if (root->val > key) {
      root->left = deleteNode(root->left, key);
    } else if (root->val < key) {
      root->right = deleteNode(root->right, key);
    } else {
      if (!root->left) {
        TreeNode* new_root = root->right;
        delete root;
        return new_root;
      } else if (!root->right) {
        TreeNode* new_root = root->left;
        delete root;
        return new_root;
      } else {
        TreeNode* node = findMinNode(root->right);
        root->val = node->val;
        root->right = deleteNode(root->right, root->val);
      }
    }
    return root;
  }

  TreeNode* findMinNode(TreeNode* node) {
    while (node->left) {
      node = node->left;
    }
    return node;
  }
};

int main() {
  Solution sol;
  TreeNode* root;
  TreeNode* got;
  int key;

  //     5             5
  //    / \           / \
  //   3   6    ->   4   6
  //  / \   \       /     \
  // 2   4   7     2       7
  root = CreateTreeNode("5 3 6 2 4 # 7");
  key = 3;
  got = sol.deleteNode(root, key);
  EXPECT_EQ(5, got->val);
  EXPECT_EQ(4, got->left->val);
  EXPECT_EQ(6, got->right->val);
  EXPECT_EQ(2, got->left->left->val);
  EXPECT_EQ(7, got->right->right->val);
  DeleteTreeNode(got);

  root = CreateTreeNode("5 3 6 2 4 # 7");
  key = 0;
  got = sol.deleteNode(root, key);
  EXPECT_EQ(5, got->val);
  EXPECT_EQ(3, got->left->val);
  EXPECT_EQ(6, got->right->val);
  EXPECT_EQ(2, got->left->left->val);
  EXPECT_EQ(4, got->left->right->val);
  EXPECT_EQ(7, got->right->right->val);
  DeleteTreeNode(got);

  root = CreateTreeNode("0");
  key = 0;
  got = sol.deleteNode(root, key);
  EXPECT_FALSE(got);

  //      6            6            6
  //     /            /            /
  //    3            2            4
  //  /   \   ->   /   \   or   /   \
  // 1     5      1     5      1     5
  //  \   /            /        \
  //   2 4            4          2
  root = CreateTreeNode("6 3 # 1 5 # 2 4");
  key = 3;
  got = sol.deleteNode(root, key);
  EXPECT_EQ(6, got->val);
  EXPECT_EQ(4, got->left->val);
  EXPECT_EQ(1, got->left->left->val);
  EXPECT_EQ(5, got->left->right->val);
  EXPECT_EQ(2, got->left->left->right->val);
  DeleteTreeNode(got);

  exit(EXIT_SUCCESS);
}
